src/sage/rings/polynomial/polynomial_element.pyx

```diff
@@ -2047,7 +2047,376 @@ cdef class Polynomial(CommutativePolynomial):
         else:
             return (False, None) if root else False
 
-    def any_root(self, ring=None, degree=None, assume_squarefree=False):
+    def _distinct_degree_factorisation_squarefree(self):
+        """
+        Helper function for any_irreducible_factor which computes
+        the distinct degree factorisation of `self`.
+
+        Creates an iterator for all valid degrees `d`, and return
+        tuples of the form `(a_d, d)` for a polynomial `a_d` the
+        product of irreducible polynomials of degree `d`
+
+        Assumes that self is squarefree.
+
+        EXAMPLES::
+
+            sage: # needs sage.rings.finite_rings
+            sage: F = GF(163)
+            sage: R.<x> = F[]
+            sage: f = (x + 162) * (x^3 + 7*x + 161) * (x^7 + 9*x + 161)
+            sage: list(f._distinct_degree_factorisation_squarefree())
+            [(x + 162, 1), (1, 2), (x^3 + 7*x + 161, 3), (x^7 + 9*x + 161, 7)]
+
+        ::
+
+            sage: # needs sage.rings.finite_rings
+            sage: K.<z8> = GF(2^8)
+            sage: R.<x> = K[]
+            sage: u1 = x + z8^6 + z8^3 + z8
+            sage: u2 = x^2 + (z8^5 + z8^3 + z8^2 + 1)*x + z8^4 + z8^3 + z8 + 1
+            sage: u3 = x^3 + z8^7 + z8^4 + z8^3 + z8^2 + z8
+            sage: f = u1 * u2 * u3
+            sage: list(f._distinct_degree_factorisation_squarefree())
+            [(x + z8^6 + z8^3 + z8, 1),
+            (x^2 + (z8^5 + z8^3 + z8^2 + 1)*x + z8^4 + z8^3 + z8 + 1, 2),
+            (x^3 + z8^7 + z8^4 + z8^3 + z8^2 + z8, 3)]
+        """
+        q = self.base_ring().order() # p^k
+        R, x = self.parent().objgen()
+
+        # Initialise values
+        v = self
+        w = x
+        d = 0
+        e = v.degree()
+
+        # Iterate over all possible degrees with degree
+        while 2*(d + 1) <= e:
+            d = d + 1
+            w = pow(w, q, v)
+
+            ad = v.gcd(w - x)
+            yield (ad, d)
+
+            if not ad.is_one():
+                v = v // ad
+                w = w % v
```

Comment:
I think this line is not needed: The `pow()` call in the next iteration should reduce the input modulo `v` first thing.